src/cmd/cmd.zig

```diff
@@ -4,48 +4,45 @@ const cli = @import("zig-cli");
 const dns = @import("zigdig");
 const network = @import("zig-network");
 const helpers = @import("helpers.zig");
+const sig = @import("../lib.zig");
+const config = @import("config.zig");
 
 const Atomic = std.atomic.Value;
 const KeyPair = std.crypto.sign.Ed25519.KeyPair;
 const Random = std.rand.Random;
 const Socket = network.Socket;
 
-const ContactInfo = @import("../gossip/data.zig").ContactInfo;
-const GossipService = @import("../gossip/service.zig").GossipService;
-const IpAddr = @import("../net/net.zig").IpAddr;
-const Level = @import("../trace/level.zig").Level;
-const Logger = @import("../trace/log.zig").Logger;
-const Pubkey = @import("../core/pubkey.zig").Pubkey;
-const Registry = @import("../prometheus/registry.zig").Registry;
-const RepairService = @import("../tvu/repair_service.zig").RepairService;
-const RepairPeerProvider = @import("../tvu/repair_service.zig").RepairPeerProvider;
-const RepairRequester = @import("../tvu/repair_service.zig").RepairRequester;
-const ShredReceiver = @import("../tvu/shred_receiver.zig").ShredReceiver;
-const SocketAddr = @import("../net/net.zig").SocketAddr;
-const SnapshotFiles = @import("../accountsdb/snapshots.zig").SnapshotFiles;
-const SnapshotFieldsAndPaths = @import("../accountsdb/snapshots.zig").SnapshotFieldsAndPaths;
-const AllSnapshotFields = @import("../accountsdb/snapshots.zig").AllSnapshotFields;
-const AccountsDB = @import("../accountsdb/db.zig").AccountsDB;
-const AccountsDBConfig = @import("../accountsdb/db.zig").AccountsDBConfig;
-const GenesisConfig = @import("../accountsdb/genesis_config.zig").GenesisConfig;
-const StatusCache = @import("../accountsdb/snapshots.zig").StatusCache;
-const SnapshotFields = @import("../accountsdb/snapshots.zig").SnapshotFields;
-const Bank = @import("../accountsdb/bank.zig").Bank;
-const enumFromName = @import("../utils/types.zig").enumFromName;
+const AccountsDB = sig.accounts_db.AccountsDB;
```

Comment:
I didn't know we wanted to move away from this import approach. I realize Ahmad removed it during the 0.12 upgrade. I thought this was because he was facing a ton of compiler errors, and he thought this import style might not be compatible with 0.12, so he removed it to rule that out. It turns out that there are no compatibility issues with 0.12, so I brought it back.

I agree that if we want to get rid of usingnamespace, then this import style is going to need some extra work to deal with. Either we'll need to change the imports in each file, or we'll need to add explicit exports to lib.zig. But I guess I view that as a one-time cost that is worthwhile because this import style has several advantages. To get ahead of the issue, I could eliminate usingnamespace in this PR, and fix the imports now.

The only advantage I see to file-based imports is that you don't need to fix imports everywhere when there is a fundamental restructuring of lib.zig. Outside the usingnamespace situation, It's hard to imagine many situations where we would need to fundamentally restructure lib.zig.

lib-based imports have a complementary advantage in that you don't need to change imports everywhere when files or file contents are moved around. You just need to make sure lib.zig exposes the same interface. In general, I see this as a bigger advantage because code moves around pretty often. lib imports allow you to decouple imports from the code's structure. lib.zig can specialize in providing a stable import interface, meanwhile we can refactor the rest of the codebase with minimal pain.

some other advantages of lib-based imports:
- ensures all tests are run
- easier to use fully-qualified names (enables faster prototyping)
- more consistent looking imports, easier to copy and paste
- dogfooding our public api
- more concise and readable (in my opinion)

Reply:
curious what you think about this approach: cf2360236cb59b67626ae1cdec03dd83942a8d98

Reply:
i like the general idea of having a lib.zig for each of the sub-directories which group imports but dont like the _private with explicit public exports - would ideally want everything thats public auto-accessible from anywhere in the codebase

Reply:
I'm open to changing this, but I'm just going to explain my thought process behind this because I was trying to accomplish something.

Many languages support visibility modifiers that allow you to declare something as public to only the current library. In rust you might use `pub(crate)` to express this. This reduces the unnecessary promises that your code needs to satisfy in the long term. It's easier to refactor, and to reason about the code's responsibilities, if it has a simple, well-defined API. In Zig, a declaration is either private to a file, or public to the entire world, so it's hard to accomplish this.

By creating all these lib.zig files, I defined each folder as its own library. The lib.zig file defines how you can import things from the library. In this file, I distinguished the items that are public to the whole world from the items that are public to only the library.

The way I happened to implement this was with two sections of the lib.zig. There's a "_private" section of consts that serves to define the items that should be visible from everywhere within the current library, but not necessarily exposed outside the library. Then the other section is just the top level of the file, which can declare its own pub consts. These pub consts are the actual public-facing API for the library. Technically _private is also accessible from anywhere. It's more of a suggestion or warning than a restriction.

So there are multiple questions here:
1. Should we distinguish between fully-public items and package-private items?
2. Is my current design a good approach to accomplish 1?

You might have also observed that most of my public exports are direct structs/functions/constants, and not files. There are some exceptions to this, like `gossip.data`. But typically I used the more specific approach because:
- it has better compatibility with existing imports
- lib.zig becomes a one-stop shop to identify at a high level what the library makes available
- it's more conservative: a simpler api is easier to maintain
- imports don't depend as much on file structure
- imports are more concise

Reply:
that makes sense and is a good way to achieve the pub(crate) goal

my personal opinion is still to have everything as public - many times to modify the rust codebase ive had to recompile or fork the repo just to make a struct/value which is private, public - yes its still accessible through _private, but that feels wrong ... if its public, it should just be public

i do like the concise imports though, thats nice

Reply:
32c08136dbc729e34c92234bf7209d773c1c56a4

I'm fine going with your preference here. Here is my initial change; it was the easiest way to make everything public.

The lingering dilemma is whether to keep the struct/function exports in lib.zig. They become redundant with the file-based exports. But I like the specific imports. I think all the bullet points from my previous comment still apply. But it's weird when there are multiple ways to do the same thing. A third approach would to manually create the same namespace that you would get with `usingnamespace`, but do it manually by explicitly creating a pub const for every decl in every file.

------
### Extended discussion

My usual approach is that everything should be private, unless it has a specific reason that it needs to be public. And then it should only be public in the scope where it might actually be used. This is to serve that goal of simplifying interfaces to minimize responsibilities and long-term maintenance cost. Basically, I want to avoid spaghetti code.

> many times to modify the rust codebase ive had to recompile or fork the repo just to make a struct/value which is private, public

I agree this is a cost, but I see it as a tradeoff. Library users need to do more work in the short term, but it can save effort for both library maintainers and library users in the long term. It's a healthy tradeoff iff there's a net reduction in effort by imposing the constraint.

I can see an argument that greenfield development is easier when things are more open. Also, the only user of these libraries that we care about right now is just other sig code in the same repo, so we're not actually taking on a huge burden here. Maybe in the long term it will become evident when it's beneficial to establish tighter constraints.

> yes its still accessible through _private, but that feels wrong ... if its public, it should just be public

Just to be clear, I only made it pub out of necessity, not because I wanted the loophole. There are other approaches to accomplish package-privacy that don't use a "private" declaration in lib.zig, but technically they can all be circumvented.

I get that you don't like the inconsistency where something *feels* private but technically *is* public. Actually you could argue that the same is true of all visibility modifiers in any language. There is always a workaround. My though process is that if you're sure that something definitely *should* be package private, it's better to have a solution which is internally inconsistent but at least partially consistent with your values, rather than having a solution that is internally consistent but fully inconsistent with your values.

> my personal opinion is still to have everything as public - many times to modify the rust codebase ive had to recompile or fork the repo just to make a struct/value which is private, public.

I feel like this argument applies equally to file-privacy and package-privacy. Zig does support file-privacy, and we take advantage of it. But if I follow this line of reasoning, I think every single declaration should be pub.

Reply:
thats very well explained - thanks for writing up your thoughts

> a struct should never have a private method that's only used internally.

i think this is the case for 99% of the time - in certain cases you dont want to expose methods because when used incorrectly they would break something so it makes sense to make them private (eg, we dont want people to access the private field in the Mux() struct because that defeats the whole purpose of the struct)

but overall, the way you/we are using the library is not going to be the same as someone else will - making things private to ensure they are used in our use case limits other use cases - most 'other use cases' will likely be things we wouldnt even think of and so imo its best to keep everything open and public

> So I usually make everything private unless it has a specific reason that I need to expose it to a particular scope. This is to serve that goal of simplifying interfaces to minimize responsibilities and long-term maintenance cost.

I dont fully understand the additional maintainers burden you describe of having more things public - what additional responsibilities are you referring to?